// @tatumio/tatum@4.2.43/dist/src/service/notification/notification.dto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotificationType = void 0;
var NotificationType;
(function (NotificationType) {
    NotificationType["ADDRESS_EVENT"] = "ADDRESS_EVENT";
    NotificationType["INCOMING_NATIVE_TX"] = "INCOMING_NATIVE_TX";
    NotificationType["OUTGOING_NATIVE_TX"] = "OUTGOING_NATIVE_TX";
    NotificationType["OUTGOING_FAILED_TX"] = "OUTGOING_FAILED_TX";
    NotificationType["PAID_FEE"] = "PAID_FEE";
    NotificationType["INCOMING_INTERNAL_TX"] = "INCOMING_INTERNAL_TX";
    NotificationType["OUTGOING_INTERNAL_TX"] = "OUTGOING_INTERNAL_TX";
    NotificationType["INCOMING_FUNGIBLE_TX"] = "INCOMING_FUNGIBLE_TX";
    NotificationType["OUTGOING_FUNGIBLE_TX"] = "OUTGOING_FUNGIBLE_TX";
    NotificationType["INCOMING_NFT_TX"] = "INCOMING_NFT_TX";
    NotificationType["OUTGOING_NFT_TX"] = "OUTGOING_NFT_TX";
    NotificationType["INCOMING_MULTITOKEN_TX"] = "INCOMING_MULTITOKEN_TX";
    NotificationType["OUTGOING_MULTITOKEN_TX"] = "OUTGOING_MULTITOKEN_TX";
    NotificationType["CONTRACT_ADDRESS_LOG_EVENT"] = "CONTRACT_ADDRESS_LOG_EVENT";
    NotificationType["FAILED_TXS_PER_BLOCK"] = "FAILED_TXS_PER_BLOCK";
})(NotificationType || (exports.NotificationType = NotificationType = {}));
