// @injectivelabs/sdk-ts/dist/cjs/client/chain/grpc/ChainGrpcPermissionsApi.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainGrpcPermissionsApi = void 0;
const exceptions_1 = require("@injectivelabs/exceptions");
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
const BaseGrpcConsumer_1 = __importDefault(require("../../base/BaseGrpcConsumer"));
const transformers_1 = require("../transformers");
const types_1 = require("../types");
/**
 * @category Chain Grpc API
 */
class ChainGrpcPermissionsApi extends BaseGrpcConsumer_1.default {
    constructor(endpoint) {
        super(endpoint);
        this.module = types_1.ChainModule.Permissions;
        this.client = new core_proto_ts_1.InjectivePermissionsV1Beta1Query.QueryClientImpl(this.getGrpcWebImpl(endpoint));
    }
    fetchAddressesByRole(_a) {
        return __awaiter(this, arguments, void 0, function* ({ denom, role }) {
            const request = core_proto_ts_1.InjectivePermissionsV1Beta1Query.QueryAddressesByRoleRequest.create();
            request.denom = denom;
            request.role = role;
            try {
                const response = yield this.retry(() => this.client.AddressesByRole(request, this.metadata));
                return transformers_1.ChainGrpcPermissionsTransformer.addressesByRolesResponseToAddressesByRoles(response);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.InjectivePermissionsV1Beta1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        context: 'AddressesByRole',
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    context: 'AddressesByRole',
                });
            }
        });
    }
    fetchAddressRoles(_a) {
        return __awaiter(this, arguments, void 0, function* ({ address, denom, }) {
            const request = core_proto_ts_1.InjectivePermissionsV1Beta1Query.QueryAddressRolesRequest.create();
            request.address = address;
            request.denom = denom;
            try {
                const response = yield this.retry(() => this.client.AddressRoles(request, this.metadata));
                return transformers_1.ChainGrpcPermissionsTransformer.addressRolesResponseToAddressRoles(response);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.InjectivePermissionsV1Beta1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        context: 'AddressRoles',
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    context: 'AddressRoles',
                });
            }
        });
    }
    fetchAllNamespaces() {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.InjectivePermissionsV1Beta1Query.QueryAllNamespacesRequest.create();
            try {
                const response = yield this.retry(() => this.client.AllNamespaces(request, this.metadata));
                return transformers_1.ChainGrpcPermissionsTransformer.allNamespacesResponseToAllNamespaces(response);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.InjectivePermissionsV1Beta1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        context: 'AllNamespaces',
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    context: 'AllNamespaces',
                });
            }
        });
    }
    fetchModuleParams() {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.InjectivePermissionsV1Beta1Query.QueryParamsRequest.create();
            try {
                const response = yield this.retry(() => this.client.Params(request, this.metadata));
                return transformers_1.ChainGrpcPermissionsTransformer.moduleParamsResponseToModuleParams(response);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.InjectivePermissionsV1Beta1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        context: 'Params',
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    context: 'Params',
                });
            }
        });
    }
    fetchNamespaceByDenom(_a) {
        return __awaiter(this, arguments, void 0, function* ({ denom, includeRoles, }) {
            const request = core_proto_ts_1.InjectivePermissionsV1Beta1Query.QueryNamespaceByDenomRequest.create();
            request.denom = denom;
            request.includeRoles = includeRoles;
            try {
                const response = yield this.retry(() => this.client.NamespaceByDenom(request, this.metadata));
                return transformers_1.ChainGrpcPermissionsTransformer.namespaceByDenomResponceToNamespaceByDenom(response);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.InjectivePermissionsV1Beta1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        context: 'NamespaceByDenom',
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    context: 'NamespaceByDenom',
                });
            }
        });
    }
    fetchVouchersForAddress(_a) {
        return __awaiter(this, arguments, void 0, function* ({ address }) {
            const request = core_proto_ts_1.InjectivePermissionsV1Beta1Query.QueryVouchersForAddressRequest.create();
            request.address = address;
            try {
                const response = yield this.retry(() => this.client.VouchersForAddress(request, this.metadata));
                return transformers_1.ChainGrpcPermissionsTransformer.vouchersForAddressResponseToVouchersForAddress(response);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.InjectivePermissionsV1Beta1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        context: 'VouchersForAddress',
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    context: 'VouchersForAddress',
                });
            }
        });
    }
}
exports.ChainGrpcPermissionsApi = ChainGrpcPermissionsApi;
