// @injectivelabs/sdk-ts@1.14.16/dist/cjs/core/modules/permissions/msgs/MsgUpdateNamespace.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const MsgBase_1 = require("../../MsgBase");
const snakecase_keys_1 = __importDefault(require("snakecase-keys"));
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
/**
 * @category Messages
 */
class MsgUpdateNamespace extends MsgBase_1.MsgBase {
    static fromJSON(params) {
        return new MsgUpdateNamespace(params);
    }
    toProto() {
        const { params } = this;
        const message = core_proto_ts_1.InjectivePermissionsV1Beta1Tx.MsgUpdateNamespace.create();
        message.sender = params.sender;
        message.namespaceDenom = params.namespaceDenom;
        message.wasmHook = {
            newValue: params.wasmHook.newValue,
        };
        message.mintsPaused = {
            newValue: params.mintsPaused.newValue,
        };
        message.sendsPaused = {
            newValue: params.sendsPaused.newValue,
        };
        message.burnsPaused = {
            newValue: params.burnsPaused.newValue,
        };
        return core_proto_ts_1.InjectivePermissionsV1Beta1Tx.MsgUpdateNamespace.fromPartial(message);
    }
    toData() {
        const proto = this.toProto();
        return Object.assign({ '@type': '/injective.permissions.v1beta1.MsgUpdateNamespace' }, proto);
    }
    toAmino() {
        const proto = this.toProto();
        const message = Object.assign({}, (0, snakecase_keys_1.default)(proto));
        return {
            type: 'permissions/MsgUpdateNamespace',
            value: message,
        };
    }
    toWeb3() {
        const amino = this.toAmino();
        const { value } = amino;
        return Object.assign({ '@type': '/injective.permissions.v1beta1.MsgUpdateNamespace' }, value);
    }
    toDirectSign() {
        const proto = this.toProto();
        return {
            type: '/injective.permissions.v1beta1.MsgUpdateNamespace',
            message: proto,
        };
    }
    toBinary() {
        return core_proto_ts_1.InjectivePermissionsV1Beta1Tx.MsgUpdateNamespace.encode(this.toProto()).finish();
    }
}
exports.default = MsgUpdateNamespace;
