// @injectivelabs/sdk-ts/dist/esm/core/modules/gov/msgs/MsgSubmitProposalPerpetualMarketLaunch.js
import snakecaseKeys from 'snakecase-keys';
import { GoogleProtobufAny, CosmosGovV1Beta1Tx, CosmosBaseV1Beta1Coin, InjectiveExchangeV1Beta1Proposal, } from '@injectivelabs/core-proto-ts';
import { MsgBase } from '../../MsgBase';
import { amountToCosmosSdkDecAmount } from '../../../../utils/numbers';
const createPerpetualMarketLaunch = (params) => {
    const content = InjectiveExchangeV1Beta1Proposal.PerpetualMarketLaunchProposal.create();
    content.title = params.market.title;
    content.description = params.market.description;
    content.ticker = params.market.ticker;
    content.quoteDenom = params.market.quoteDenom;
    content.oracleBase = params.market.oracleBase;
    content.oracleQuote = params.market.oracleQuote;
    content.oracleScaleFactor = params.market.oracleScaleFactor;
    content.oracleType = params.market.oracleType;
    content.initialMarginRatio = params.market.initialMarginRatio;
    content.maintenanceMarginRatio = params.market.maintenanceMarginRatio;
    content.makerFeeRate = params.market.makerFeeRate;
    content.takerFeeRate = params.market.takerFeeRate;
    content.minPriceTickSize = params.market.minPriceTickSize;
    content.minQuantityTickSize = params.market.minQuantityTickSize;
    content.minNotional = params.market.minNotional;
    return InjectiveExchangeV1Beta1Proposal.PerpetualMarketLaunchProposal.fromPartial(content);
};
/**
 * @category Messages
 */
export default class MsgSubmitProposalPerpetualMarketLaunch extends MsgBase {
    static fromJSON(params) {
        return new MsgSubmitProposalPerpetualMarketLaunch(params);
    }
    toProto() {
        const { params } = this;
        const depositParams = CosmosBaseV1Beta1Coin.Coin.create();
        depositParams.denom = params.deposit.denom;
        depositParams.amount = params.deposit.amount;
        const content = createPerpetualMarketLaunch({
            ...params,
            market: {
                ...params.market,
                initialMarginRatio: amountToCosmosSdkDecAmount(params.market.initialMarginRatio).toFixed(),
                maintenanceMarginRatio: amountToCosmosSdkDecAmount(params.market.maintenanceMarginRatio).toFixed(),
                makerFeeRate: amountToCosmosSdkDecAmount(params.market.makerFeeRate).toFixed(),
                takerFeeRate: amountToCosmosSdkDecAmount(params.market.takerFeeRate).toFixed(),
                minQuantityTickSize: amountToCosmosSdkDecAmount(params.market.minQuantityTickSize).toFixed(),
                minNotional: amountToCosmosSdkDecAmount(params.market.minNotional).toFixed(),
            },
        });
        const contentAny = GoogleProtobufAny.Any.create();
        contentAny.typeUrl =
            '/injective.exchange.v1beta1.PerpetualMarketLaunchProposal';
        contentAny.value =
            InjectiveExchangeV1Beta1Proposal.PerpetualMarketLaunchProposal.encode(content).finish();
        const message = CosmosGovV1Beta1Tx.MsgSubmitProposal.create();
        message.content = contentAny;
        message.initialDeposit = [depositParams];
        message.proposer = params.proposer;
        return CosmosGovV1Beta1Tx.MsgSubmitProposal.fromPartial(message);
    }
    toData() {
        const proto = this.toProto();
        return {
            '@type': '/cosmos.gov.v1beta1.MsgSubmitProposal',
            ...proto,
        };
    }
    toAmino() {
        const { params } = this;
        const content = this.getContent();
        const message = {
            content,
            proposer: params.proposer,
        };
        const messageWithProposalType = snakecaseKeys({
            content: {
                type: 'exchange/PerpetualMarketLaunchProposal',
                value: message.content,
            },
            initialDeposit: [
                {
                    denom: params.deposit.denom,
                    amount: params.deposit.amount,
                },
            ],
            proposer: params.proposer,
        });
        return {
            type: 'cosmos-sdk/MsgSubmitProposal',
            value: messageWithProposalType,
        };
    }
    toWeb3() {
        const { params } = this;
        const messageWithProposalType = {
            content: {
                '@type': '/injective.exchange.v1beta1.PerpetualMarketLaunchProposal',
                ...this.getContent(),
            },
            initialDeposit: [
                {
                    denom: params.deposit.denom,
                    amount: params.deposit.amount,
                },
            ],
            proposer: params.proposer,
        };
        return {
            '@type': '/cosmos.gov.v1beta1.MsgSubmitProposal',
            ...messageWithProposalType,
        };
    }
    toDirectSign() {
        const proto = this.toProto();
        return {
            type: '/cosmos.gov.v1beta1.MsgSubmitProposal',
            message: proto,
        };
    }
    toBinary() {
        return CosmosGovV1Beta1Tx.MsgSubmitProposal.encode(this.toProto()).finish();
    }
    getContent() {
        const { params } = this;
        return createPerpetualMarketLaunch(params);
    }
}
