// @injectivelabs/core-proto-ts@0.0.34/cjs/cometbft/services/block_results/v1/block_results_service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcWebError = exports.GrpcWebImpl = exports.BlockResultsServiceGetBlockResultsDesc = exports.BlockResultsServiceDesc = exports.BlockResultsServiceClientImpl = exports.protobufPackage = void 0;
/* eslint-disable */
const grpc_web_1 = require("@injectivelabs/grpc-web");
const browser_headers_1 = require("browser-headers");
const block_results_1 = require("./block_results");
exports.protobufPackage = "cometbft.services.block_results.v1";
class BlockResultsServiceClientImpl {
    constructor(rpc) {
        this.rpc = rpc;
        this.GetBlockResults = this.GetBlockResults.bind(this);
    }
    GetBlockResults(request, metadata) {
        return this.rpc.unary(exports.BlockResultsServiceGetBlockResultsDesc, block_results_1.GetBlockResultsRequest.fromPartial(request), metadata);
    }
}
exports.BlockResultsServiceClientImpl = BlockResultsServiceClientImpl;
exports.BlockResultsServiceDesc = { serviceName: "cometbft.services.block_results.v1.BlockResultsService" };
exports.BlockResultsServiceGetBlockResultsDesc = {
    methodName: "GetBlockResults",
    service: exports.BlockResultsServiceDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return block_results_1.GetBlockResultsRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            const value = block_results_1.GetBlockResultsResponse.decode(data);
            return Object.assign(Object.assign({}, value), { toObject() {
                    return value;
                } });
        },
    },
};
class GrpcWebImpl {
    constructor(host, options) {
        this.host = host;
        this.options = options;
    }
    unary(methodDesc, _request, metadata) {
        var _a;
        const request = Object.assign(Object.assign({}, _request), methodDesc.requestType);
        const maybeCombinedMetadata = metadata && this.options.metadata
            ? new browser_headers_1.BrowserHeaders(Object.assign(Object.assign({}, (_a = this.options) === null || _a === void 0 ? void 0 : _a.metadata.headersMap), metadata === null || metadata === void 0 ? void 0 : metadata.headersMap))
            : metadata || this.options.metadata;
        return new Promise((resolve, reject) => {
            grpc_web_1.grpc.unary(methodDesc, {
                request,
                host: this.host,
                metadata: maybeCombinedMetadata,
                transport: this.options.transport,
                debug: this.options.debug,
                onEnd: function (response) {
                    if (response.status === grpc_web_1.grpc.Code.OK) {
                        resolve(response.message.toObject());
                    }
                    else {
                        const err = new GrpcWebError(response.statusMessage, response.status, response.trailers);
                        reject(err);
                    }
                },
            });
        });
    }
}
exports.GrpcWebImpl = GrpcWebImpl;
var tsProtoGlobalThis = (() => {
    if (typeof globalThis !== "undefined") {
        return globalThis;
    }
    if (typeof self !== "undefined") {
        return self;
    }
    if (typeof window !== "undefined") {
        return window;
    }
    if (typeof global !== "undefined") {
        return global;
    }
    throw "Unable to locate global object";
})();
class GrpcWebError extends tsProtoGlobalThis.Error {
    constructor(message, code, metadata) {
        super(message);
        this.code = code;
        this.metadata = metadata;
    }
}
exports.GrpcWebError = GrpcWebError;
