// @ensdomains/ens-validation@0.1.0/dist/lib/enums/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpoofChecks = exports.RestrictionLevel = exports.ErrorCode = void 0;
var ErrorCode;
(function (ErrorCode) {
    ErrorCode[ErrorCode["ZERO_ERROR"] = 0] = "ZERO_ERROR";
    ErrorCode[ErrorCode["SUCCESS"] = 1] = "SUCCESS";
    ErrorCode[ErrorCode["FAILURE"] = 2] = "FAILURE";
    ErrorCode[ErrorCode["INVALID_FORMAT_ERROR"] = 3] = "INVALID_FORMAT_ERROR";
    ErrorCode[ErrorCode["ILLEGAL_ARGUMENT_ERROR"] = 4] = "ILLEGAL_ARGUMENT_ERROR";
    ErrorCode[ErrorCode["MEMORY_ALLOCATION_ERROR"] = 5] = "MEMORY_ALLOCATION_ERROR";
})(ErrorCode = exports.ErrorCode || (exports.ErrorCode = {}));
var RestrictionLevel;
(function (RestrictionLevel) {
    /**
     * All characters in the string are in the identifier profile and all characters in the string are in the
     * ASCII range.
     *
     * @stable ICU 51
     */
    RestrictionLevel[RestrictionLevel["ASCII"] = 268435456] = "ASCII";
    /**
     * The string classifies as ASCII-Only, or all characters in the string are in the identifier profile and
     * the string is single-script, according to the definition in UTS 39 section 5.1.
     *
     * @stable ICU 53
     */
    RestrictionLevel[RestrictionLevel["SINGLE_SCRIPT_RESTRICTIVE"] = 536870912] = "SINGLE_SCRIPT_RESTRICTIVE";
    /**
     * The string classifies as Single Script, or all characters in the string are in the identifier profile and
     * the string is covered by any of the following sets of scripts, according to the definition in UTS 39
     * section 5.1:
     * <ul>
     *   <li>Latin + Han + Bopomofo (or equivalently: Latn + Hanb)</li>
     *   <li>Latin + Han + Hiragana + Katakana (or equivalently: Latn + Jpan)</li>
     *   <li>Latin + Han + Hangul (or equivalently: Latn +Kore)</li>
     * </ul>
     * This is the default restriction in ICU.
     *
     * @stable ICU 51
     */
    RestrictionLevel[RestrictionLevel["HIGHLY_RESTRICTIVE"] = 805306368] = "HIGHLY_RESTRICTIVE";
    /**
     * The string classifies as Highly Restrictive, or all characters in the string are in the identifier profile
     * and the string is covered by Latin and any one other Recommended or Aspirational script, except Cyrillic,
     * Greek, and Cherokee.
     *
     * @stable ICU 51
     */
    RestrictionLevel[RestrictionLevel["MODERATELY_RESTRICTIVE"] = 1073741824] = "MODERATELY_RESTRICTIVE";
    /**
     * All characters in the string are in the identifier profile.  Allow arbitrary mixtures of scripts.
     *
     * @stable ICU 51
     */
    RestrictionLevel[RestrictionLevel["MINIMALLY_RESTRICTIVE"] = 1342177280] = "MINIMALLY_RESTRICTIVE";
    /**
     * Any valid identifiers, including characters outside of the Identifier Profile.
     *
     * @stable ICU 51
     */
    RestrictionLevel[RestrictionLevel["UNRESTRICTIVE"] = 1610612736] = "UNRESTRICTIVE";
    /**
     * Mask for selecting the Restriction Level bits from the return value of {@link uspoof_check}.
     *
     * @stable ICU 53
     */
    RestrictionLevel[RestrictionLevel["RESTRICTION_LEVEL_MASK"] = 2130706432] = "RESTRICTION_LEVEL_MASK";
    /**
     * An undefined restriction level.
     * @internal
     */
    RestrictionLevel[RestrictionLevel["UNDEFINED_RESTRICTIVE"] = -1] = "UNDEFINED_RESTRICTIVE";
})(RestrictionLevel = exports.RestrictionLevel || (exports.RestrictionLevel = {}));
var SpoofChecks;
(function (SpoofChecks) {
    SpoofChecks[SpoofChecks["SINGLE_SCRIPT_CONFUSABLE"] = 1] = "SINGLE_SCRIPT_CONFUSABLE";
    SpoofChecks[SpoofChecks["MIXED_SCRIPT_CONFUSABLE"] = 2] = "MIXED_SCRIPT_CONFUSABLE";
    SpoofChecks[SpoofChecks["WHOLE_SCRIPT_CONFUSABLE"] = 4] = "WHOLE_SCRIPT_CONFUSABLE";
    SpoofChecks[SpoofChecks["CONFUSABLE"] = 7] = "CONFUSABLE";
    SpoofChecks[SpoofChecks["RESTRICTION_LEVEL"] = 16] = "RESTRICTION_LEVEL";
    SpoofChecks[SpoofChecks["INVISIBLE"] = 32] = "INVISIBLE";
    SpoofChecks[SpoofChecks["CHAR_LIMIT"] = 64] = "CHAR_LIMIT";
    SpoofChecks[SpoofChecks["MIXED_NUMBERS"] = 128] = "MIXED_NUMBERS";
    SpoofChecks[SpoofChecks["ALL_CHECKS"] = 65535] = "ALL_CHECKS";
    // AUX_INFO = 0x40000000,
})(SpoofChecks = exports.SpoofChecks || (exports.SpoofChecks = {}));
