// @injectivelabs/sdk-ts/dist/cjs/core/modules/permissions/msgs/MsgRevokeNamespaceRoles.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const MsgBase_1 = require("../../MsgBase");
const snakecase_keys_1 = __importDefault(require("snakecase-keys"));
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
/**
 * @category Messages
 */
class MsgRevokeNamespaceRoles extends MsgBase_1.MsgBase {
    static fromJSON(params) {
        return new MsgRevokeNamespaceRoles(params);
    }
    toProto() {
        const { params } = this;
        const message = core_proto_ts_1.InjectivePermissionsV1Beta1Tx.MsgRevokeNamespaceRoles.create();
        message.sender = params.sender;
        message.namespaceDenom = params.namespaceDenom;
        message.addressRolesToRevoke = params.addressRolesToRevoke;
        return core_proto_ts_1.InjectivePermissionsV1Beta1Tx.MsgRevokeNamespaceRoles.fromPartial(message);
    }
    toData() {
        const proto = this.toProto();
        return Object.assign({ '@type': '/injective.permissions.v1beta1.MsgRevokeNamespaceRoles' }, proto);
    }
    toAmino() {
        const proto = this.toProto();
        const message = Object.assign({}, (0, snakecase_keys_1.default)(proto));
        return {
            type: 'permissions/MsgRevokeNamespaceRoles',
            value: message,
        };
    }
    toWeb3() {
        const amino = this.toAmino();
        const { value } = amino;
        return Object.assign({ '@type': '/injective.permissions.v1beta1.MsgRevokeNamespaceRoles' }, value);
    }
    toDirectSign() {
        const proto = this.toProto();
        return {
            type: '/injective.permissions.v1beta1.MsgRevokeNamespaceRoles',
            message: proto,
        };
    }
    toBinary() {
        return core_proto_ts_1.InjectivePermissionsV1Beta1Tx.MsgRevokeNamespaceRoles.encode(this.toProto()).finish();
    }
}
exports.default = MsgRevokeNamespaceRoles;
