// @sei-js/proto@3.1.0/dist/types/codegen/seiprotocol/seichain/mint/v1beta1/query.js
import { Params } from "./mint";
import { Long } from "../../../../helpers";
import * as _m0 from "protobufjs/minimal";
function createBaseQueryParamsRequest() {
    return {};
}
export const QueryParamsRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryParamsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseQueryParamsRequest();
        return message;
    }
};
function createBaseQueryParamsResponse() {
    return {
        params: Params.fromPartial({})
    };
}
export const QueryParamsResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.params !== undefined) {
            Params.encode(message.params, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryParamsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.params = Params.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryParamsResponse();
        message.params = object.params !== undefined && object.params !== null ? Params.fromPartial(object.params) : undefined;
        return message;
    }
};
function createBaseQueryMinterRequest() {
    return {};
}
export const QueryMinterRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryMinterRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseQueryMinterRequest();
        return message;
    }
};
function createBaseQueryMinterResponse() {
    return {
        startDate: "",
        endDate: "",
        denom: "",
        totalMintAmount: Long.UZERO,
        remainingMintAmount: Long.UZERO,
        lastMintAmount: Long.UZERO,
        lastMintDate: "",
        lastMintHeight: Long.UZERO
    };
}
export const QueryMinterResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.startDate !== "") {
            writer.uint32(10).string(message.startDate);
        }
        if (message.endDate !== "") {
            writer.uint32(18).string(message.endDate);
        }
        if (message.denom !== "") {
            writer.uint32(26).string(message.denom);
        }
        if (!message.totalMintAmount.isZero()) {
            writer.uint32(32).uint64(message.totalMintAmount);
        }
        if (!message.remainingMintAmount.isZero()) {
            writer.uint32(40).uint64(message.remainingMintAmount);
        }
        if (!message.lastMintAmount.isZero()) {
            writer.uint32(48).uint64(message.lastMintAmount);
        }
        if (message.lastMintDate !== "") {
            writer.uint32(58).string(message.lastMintDate);
        }
        if (!message.lastMintHeight.isZero()) {
            writer.uint32(64).uint64(message.lastMintHeight);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryMinterResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.startDate = reader.string();
                    break;
                case 2:
                    message.endDate = reader.string();
                    break;
                case 3:
                    message.denom = reader.string();
                    break;
                case 4:
                    message.totalMintAmount = reader.uint64();
                    break;
                case 5:
                    message.remainingMintAmount = reader.uint64();
                    break;
                case 6:
                    message.lastMintAmount = reader.uint64();
                    break;
                case 7:
                    message.lastMintDate = reader.string();
                    break;
                case 8:
                    message.lastMintHeight = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryMinterResponse();
        message.startDate = object.startDate ?? "";
        message.endDate = object.endDate ?? "";
        message.denom = object.denom ?? "";
        message.totalMintAmount = object.totalMintAmount !== undefined && object.totalMintAmount !== null ? Long.fromValue(object.totalMintAmount) : Long.UZERO;
        message.remainingMintAmount = object.remainingMintAmount !== undefined && object.remainingMintAmount !== null ? Long.fromValue(object.remainingMintAmount) : Long.UZERO;
        message.lastMintAmount = object.lastMintAmount !== undefined && object.lastMintAmount !== null ? Long.fromValue(object.lastMintAmount) : Long.UZERO;
        message.lastMintDate = object.lastMintDate ?? "";
        message.lastMintHeight = object.lastMintHeight !== undefined && object.lastMintHeight !== null ? Long.fromValue(object.lastMintHeight) : Long.UZERO;
        return message;
    }
};
