// @injectivelabs/sdk-ts/dist/esm/core/modules/staking/msgs/MsgCreateValidator.js
import { MsgBase } from '../../MsgBase';
import snakecaseKeys from 'snakecase-keys';
import { CosmosBaseV1Beta1Coin, CosmosStakingV1Beta1Tx, CosmosStakingV1Beta1Staking, } from '@injectivelabs/core-proto-ts';
import { createAny } from '../../../tx';
/**
 * @category Messages
 */
export default class MsgCreateValidator extends MsgBase {
    static fromJSON(params) {
        return new MsgCreateValidator(params);
    }
    toProto() {
        const { params } = this;
        const message = CosmosStakingV1Beta1Tx.MsgCreateValidator.create();
        if (params.description) {
            const description = CosmosStakingV1Beta1Staking.Description.create();
            if (params.description.moniker) {
                description.moniker = params.description.moniker;
            }
            if (params.description.identity) {
                description.identity = params.description.identity;
            }
            if (params.description.website) {
                description.website = params.description.website;
            }
            if (params.description.securityContact) {
                description.securityContact = params.description.securityContact;
            }
            if (params.description.details) {
                description.details = params.description.details;
            }
            message.description = description;
        }
        if (params.commission) {
            const commissionRate = CosmosStakingV1Beta1Staking.CommissionRates.create();
            commissionRate.rate = commissionRate.rate;
            commissionRate.maxRate = commissionRate.maxRate;
            commissionRate.maxChangeRate = commissionRate.maxChangeRate;
        }
        if (params.minSelfDelegation) {
            message.minSelfDelegation = params.minSelfDelegation;
        }
        message.delegatorAddress = params.delegatorAddress;
        message.validatorAddress = params.validatorAddress;
        if (params.pubKey) {
            const pubKeyAny = createAny(Buffer.from(params.pubKey.value, 'base64'), params.pubKey.type);
            message.pubkey = pubKeyAny;
        }
        if (params.value) {
            const coin = CosmosBaseV1Beta1Coin.Coin.create();
            coin.denom = params.value.denom;
            coin.amount = params.value.amount;
            message.value = coin;
        }
        return CosmosStakingV1Beta1Tx.MsgCreateValidator.fromPartial(message);
    }
    toData() {
        const proto = this.toProto();
        return {
            '@type': '/cosmos.staking.v1beta1.MsgCreateValidator',
            ...proto,
        };
    }
    toAmino() {
        const proto = this.toProto();
        const message = {
            ...snakecaseKeys(proto),
        };
        return {
            type: 'cosmos-sdk/MsgCreateValidator',
            value: message,
        };
    }
    toWeb3() {
        const amino = this.toAmino();
        const { value } = amino;
        return {
            '@type': '/cosmos.staking.v1beta1.MsgCreateValidator',
            ...value,
        };
    }
    toDirectSign() {
        const proto = this.toProto();
        return {
            type: '/cosmos.staking.v1beta1.MsgCreateValidator',
            message: proto,
        };
    }
    toBinary() {
        return CosmosStakingV1Beta1Tx.MsgCreateValidator.encode(this.toProto()).finish();
    }
}
