// @sei-js/proto/dist/esm/codegen/seiprotocol/seichain/tokenfactory/tx.registry.js
//@ts-nocheck

import { MsgCreateDenom, MsgMint, MsgBurn, MsgChangeAdmin } from "./tx";
export const registry = [["/seiprotocol.seichain.tokenfactory.MsgCreateDenom", MsgCreateDenom], ["/seiprotocol.seichain.tokenfactory.MsgMint", MsgMint], ["/seiprotocol.seichain.tokenfactory.MsgBurn", MsgBurn], ["/seiprotocol.seichain.tokenfactory.MsgChangeAdmin", MsgChangeAdmin]];
export const load = protoRegistry => {
  registry.forEach(([typeUrl, mod]) => {
    protoRegistry.register(typeUrl, mod);
  });
};
export const MessageComposer = {
  encoded: {
    createDenom(value) {
      return {
        typeUrl: "/seiprotocol.seichain.tokenfactory.MsgCreateDenom",
        value: MsgCreateDenom.encode(value).finish()
      };
    },
    mint(value) {
      return {
        typeUrl: "/seiprotocol.seichain.tokenfactory.MsgMint",
        value: MsgMint.encode(value).finish()
      };
    },
    burn(value) {
      return {
        typeUrl: "/seiprotocol.seichain.tokenfactory.MsgBurn",
        value: MsgBurn.encode(value).finish()
      };
    },
    changeAdmin(value) {
      return {
        typeUrl: "/seiprotocol.seichain.tokenfactory.MsgChangeAdmin",
        value: MsgChangeAdmin.encode(value).finish()
      };
    }
  },
  withTypeUrl: {
    createDenom(value) {
      return {
        typeUrl: "/seiprotocol.seichain.tokenfactory.MsgCreateDenom",
        value
      };
    },
    mint(value) {
      return {
        typeUrl: "/seiprotocol.seichain.tokenfactory.MsgMint",
        value
      };
    },
    burn(value) {
      return {
        typeUrl: "/seiprotocol.seichain.tokenfactory.MsgBurn",
        value
      };
    },
    changeAdmin(value) {
      return {
        typeUrl: "/seiprotocol.seichain.tokenfactory.MsgChangeAdmin",
        value
      };
    }
  },
  fromPartial: {
    createDenom(value) {
      return {
        typeUrl: "/seiprotocol.seichain.tokenfactory.MsgCreateDenom",
        value: MsgCreateDenom.fromPartial(value)
      };
    },
    mint(value) {
      return {
        typeUrl: "/seiprotocol.seichain.tokenfactory.MsgMint",
        value: MsgMint.fromPartial(value)
      };
    },
    burn(value) {
      return {
        typeUrl: "/seiprotocol.seichain.tokenfactory.MsgBurn",
        value: MsgBurn.fromPartial(value)
      };
    },
    changeAdmin(value) {
      return {
        typeUrl: "/seiprotocol.seichain.tokenfactory.MsgChangeAdmin",
        value: MsgChangeAdmin.fromPartial(value)
      };
    }
  }
};
