// @injectivelabs/sdk-ts@1.14.16/dist/esm/core/modules/permissions/msgs/MsgDeleteNamespace.js
import { MsgBase } from '../../MsgBase';
import snakecaseKeys from 'snakecase-keys';
import { InjectivePermissionsV1Beta1Tx } from '@injectivelabs/core-proto-ts';
/**
 * @category Messages
 */
export default class MsgDeleteNamespace extends MsgBase {
    static fromJSON(params) {
        return new MsgDeleteNamespace(params);
    }
    toProto() {
        const { params } = this;
        const message = InjectivePermissionsV1Beta1Tx.MsgDeleteNamespace.create();
        message.sender = params.sender;
        message.namespaceDenom = params.namespaceDenom;
        return InjectivePermissionsV1Beta1Tx.MsgDeleteNamespace.fromPartial(message);
    }
    toData() {
        const proto = this.toProto();
        return {
            '@type': '/injective.permissions.v1beta1.MsgDeleteNamespace',
            ...proto,
        };
    }
    toAmino() {
        const proto = this.toProto();
        const message = {
            ...snakecaseKeys(proto),
        };
        return {
            type: 'permissions/MsgDeleteNamespace',
            value: message,
        };
    }
    toWeb3() {
        const amino = this.toAmino();
        const { value } = amino;
        return {
            '@type': '/injective.permissions.v1beta1.MsgDeleteNamespace',
            ...value,
        };
    }
    toDirectSign() {
        const proto = this.toProto();
        return {
            type: '/injective.permissions.v1beta1.MsgDeleteNamespace',
            message: proto,
        };
    }
    toBinary() {
        return InjectivePermissionsV1Beta1Tx.MsgDeleteNamespace.encode(this.toProto()).finish();
    }
}
