// @injectivelabs/sdk-ts@1.14.16/dist/esm/core/modules/permissions/msgs/MsgUpdateNamespaceRoles.js
import { MsgBase } from '../../MsgBase';
import snakecaseKeys from 'snakecase-keys';
import { InjectivePermissionsV1Beta1Tx } from '@injectivelabs/core-proto-ts';
/**
 * @category Messages
 */
export default class MsgUpdateNamespaceRoles extends MsgBase {
    static fromJSON(params) {
        return new MsgUpdateNamespaceRoles(params);
    }
    toProto() {
        const { params } = this;
        const message = InjectivePermissionsV1Beta1Tx.MsgUpdateNamespaceRoles.create();
        message.sender = params.sender;
        message.namespaceDenom = params.namespaceDenom;
        message.rolePermissions = params.rolePermissions;
        message.addressRoles = params.addressRoles;
        return InjectivePermissionsV1Beta1Tx.MsgUpdateNamespaceRoles.fromPartial(message);
    }
    toData() {
        const proto = this.toProto();
        return {
            '@type': '/injective.permissions.v1beta1.MsgUpdateNamespaceRoles',
            ...proto,
        };
    }
    toAmino() {
        const proto = this.toProto();
        const message = {
            ...snakecaseKeys(proto),
        };
        return {
            type: 'permissions/MsgUpdateNamespaceRoles',
            value: message,
        };
    }
    toWeb3() {
        const amino = this.toAmino();
        const { value } = amino;
        return {
            '@type': '/injective.permissions.v1beta1.MsgUpdateNamespaceRoles',
            ...value,
        };
    }
    toDirectSign() {
        const proto = this.toProto();
        return {
            type: '/injective.permissions.v1beta1.MsgUpdateNamespaceRoles',
            message: proto,
        };
    }
    toBinary() {
        return InjectivePermissionsV1Beta1Tx.MsgUpdateNamespaceRoles.encode(this.toProto()).finish();
    }
}
