// @injectivelabs/core-proto-ts@0.0.34/esm/cosmos/evidence/module/v1/module.js
/* eslint-disable */
import _m0 from "protobufjs/minimal";
export const protobufPackage = "cosmos.evidence.module.v1";
function createBaseModule() {
    return {};
}
export const Module = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseModule();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return Module.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseModule();
        return message;
    },
};
