// @injectivelabs/exceptions/dist/cjs/exceptions/types/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorType = exports.HttpRequestMethod = void 0;
var HttpRequestMethod;
(function (HttpRequestMethod) {
    HttpRequestMethod["Get"] = "GET";
    HttpRequestMethod["Post"] = "POST";
    HttpRequestMethod["Options"] = "OPTIONS";
})(HttpRequestMethod || (exports.HttpRequestMethod = HttpRequestMethod = {}));
var ErrorType;
(function (ErrorType) {
    ErrorType["Unspecified"] = "unspecified";
    ErrorType["ChainError"] = "chain-error";
    ErrorType["ExecutionError"] = "execution-error";
    ErrorType["NotFoundError"] = "not-found-error";
    ErrorType["ValidationError"] = "validation-error";
    ErrorType["WalletError"] = "wallet-error";
    ErrorType["WalletNotInstalledError"] = "wallet-not-installed-error";
    ErrorType["GrpcUnaryRequest"] = "grpc-unary-request";
    ErrorType["HttpRequest"] = "http-request";
    ErrorType["Web3"] = "web3";
    ErrorType["Web3Gateway"] = "web3-gateway";
})(ErrorType || (exports.ErrorType = ErrorType = {}));
