// @sei-js/proto/dist/esm/codegen/seiprotocol/seichain/dex/gov.js
import { AssetMetadata } from "./asset_list";
import * as _m0 from "protobufjs/minimal";
function createBaseAddAssetMetadataProposal() {
  return {
    title: "",
    description: "",
    assetList: []
  };
}
export const AddAssetMetadataProposal = {
  encode(message, writer = _m0.Writer.create()) {
    if (message.title !== "") {
      writer.uint32(10).string(message.title);
    }
    if (message.description !== "") {
      writer.uint32(18).string(message.description);
    }
    for (const v of message.assetList) {
      AssetMetadata.encode(v, writer.uint32(26).fork()).ldelim();
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseAddAssetMetadataProposal();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.title = reader.string();
          break;
        case 2:
          message.description = reader.string();
          break;
        case 3:
          message.assetList.push(AssetMetadata.decode(reader, reader.uint32()));
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object) {
    var _object$title, _object$description, _object$assetList;
    const message = createBaseAddAssetMetadataProposal();
    message.title = (_object$title = object.title) !== null && _object$title !== void 0 ? _object$title : "";
    message.description = (_object$description = object.description) !== null && _object$description !== void 0 ? _object$description : "";
    message.assetList = ((_object$assetList = object.assetList) === null || _object$assetList === void 0 ? void 0 : _object$assetList.map(e => AssetMetadata.fromPartial(e))) || [];
    return message;
  }
};
