// @sei-js/proto/dist/esm/codegen/seiprotocol/seichain/dex/order.js
import { Long } from "../../../helpers";
import * as _m0 from "protobufjs/minimal";
function createBaseOrder() {
  return {
    id: undefined,
    status: undefined,
    account: undefined,
    contractAddr: undefined,
    price: "",
    quantity: "",
    priceDenom: "",
    assetDenom: "",
    orderType: 0,
    positionDirection: 0,
    data: "",
    statusDescription: "",
    nominal: "",
    triggerPrice: "",
    triggerStatus: false
  };
}
export const Order = {
  encode(message, writer = _m0.Writer.create()) {
    if (message.id !== undefined) {
      writer.uint32(8).uint64(message.id);
    }
    if (message.status !== undefined) {
      writer.uint32(16).int32(message.status);
    }
    if (message.account !== undefined) {
      writer.uint32(26).string(message.account);
    }
    if (message.contractAddr !== undefined) {
      writer.uint32(34).string(message.contractAddr);
    }
    if (message.price !== "") {
      writer.uint32(42).string(message.price);
    }
    if (message.quantity !== "") {
      writer.uint32(50).string(message.quantity);
    }
    if (message.priceDenom !== "") {
      writer.uint32(58).string(message.priceDenom);
    }
    if (message.assetDenom !== "") {
      writer.uint32(66).string(message.assetDenom);
    }
    if (message.orderType !== 0) {
      writer.uint32(72).int32(message.orderType);
    }
    if (message.positionDirection !== 0) {
      writer.uint32(80).int32(message.positionDirection);
    }
    if (message.data !== "") {
      writer.uint32(90).string(message.data);
    }
    if (message.statusDescription !== "") {
      writer.uint32(98).string(message.statusDescription);
    }
    if (message.nominal !== "") {
      writer.uint32(106).string(message.nominal);
    }
    if (message.triggerPrice !== "") {
      writer.uint32(114).string(message.triggerPrice);
    }
    if (message.triggerStatus === true) {
      writer.uint32(120).bool(message.triggerStatus);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseOrder();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.id = reader.uint64();
          break;
        case 2:
          message.status = reader.int32();
          break;
        case 3:
          message.account = reader.string();
          break;
        case 4:
          message.contractAddr = reader.string();
          break;
        case 5:
          message.price = reader.string();
          break;
        case 6:
          message.quantity = reader.string();
          break;
        case 7:
          message.priceDenom = reader.string();
          break;
        case 8:
          message.assetDenom = reader.string();
          break;
        case 9:
          message.orderType = reader.int32();
          break;
        case 10:
          message.positionDirection = reader.int32();
          break;
        case 11:
          message.data = reader.string();
          break;
        case 12:
          message.statusDescription = reader.string();
          break;
        case 13:
          message.nominal = reader.string();
          break;
        case 14:
          message.triggerPrice = reader.string();
          break;
        case 15:
          message.triggerStatus = reader.bool();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object) {
    var _object$status, _object$account, _object$contractAddr, _object$price, _object$quantity, _object$priceDenom, _object$assetDenom, _object$orderType, _object$positionDirec, _object$data, _object$statusDescrip, _object$nominal, _object$triggerPrice, _object$triggerStatus;
    const message = createBaseOrder();
    message.id = object.id !== undefined && object.id !== null ? Long.fromValue(object.id) : undefined;
    message.status = (_object$status = object.status) !== null && _object$status !== void 0 ? _object$status : undefined;
    message.account = (_object$account = object.account) !== null && _object$account !== void 0 ? _object$account : undefined;
    message.contractAddr = (_object$contractAddr = object.contractAddr) !== null && _object$contractAddr !== void 0 ? _object$contractAddr : undefined;
    message.price = (_object$price = object.price) !== null && _object$price !== void 0 ? _object$price : "";
    message.quantity = (_object$quantity = object.quantity) !== null && _object$quantity !== void 0 ? _object$quantity : "";
    message.priceDenom = (_object$priceDenom = object.priceDenom) !== null && _object$priceDenom !== void 0 ? _object$priceDenom : "";
    message.assetDenom = (_object$assetDenom = object.assetDenom) !== null && _object$assetDenom !== void 0 ? _object$assetDenom : "";
    message.orderType = (_object$orderType = object.orderType) !== null && _object$orderType !== void 0 ? _object$orderType : 0;
    message.positionDirection = (_object$positionDirec = object.positionDirection) !== null && _object$positionDirec !== void 0 ? _object$positionDirec : 0;
    message.data = (_object$data = object.data) !== null && _object$data !== void 0 ? _object$data : "";
    message.statusDescription = (_object$statusDescrip = object.statusDescription) !== null && _object$statusDescrip !== void 0 ? _object$statusDescrip : "";
    message.nominal = (_object$nominal = object.nominal) !== null && _object$nominal !== void 0 ? _object$nominal : "";
    message.triggerPrice = (_object$triggerPrice = object.triggerPrice) !== null && _object$triggerPrice !== void 0 ? _object$triggerPrice : "";
    message.triggerStatus = (_object$triggerStatus = object.triggerStatus) !== null && _object$triggerStatus !== void 0 ? _object$triggerStatus : false;
    return message;
  }
};
function createBaseCancellation() {
  return {
    id: Long.UZERO,
    initiator: 0,
    creator: undefined,
    contractAddr: "",
    priceDenom: "",
    assetDenom: "",
    positionDirection: 0,
    price: ""
  };
}
export const Cancellation = {
  encode(message, writer = _m0.Writer.create()) {
    if (!message.id.isZero()) {
      writer.uint32(8).uint64(message.id);
    }
    if (message.initiator !== 0) {
      writer.uint32(16).int32(message.initiator);
    }
    if (message.creator !== undefined) {
      writer.uint32(26).string(message.creator);
    }
    if (message.contractAddr !== "") {
      writer.uint32(34).string(message.contractAddr);
    }
    if (message.priceDenom !== "") {
      writer.uint32(42).string(message.priceDenom);
    }
    if (message.assetDenom !== "") {
      writer.uint32(50).string(message.assetDenom);
    }
    if (message.positionDirection !== 0) {
      writer.uint32(56).int32(message.positionDirection);
    }
    if (message.price !== "") {
      writer.uint32(66).string(message.price);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseCancellation();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.id = reader.uint64();
          break;
        case 2:
          message.initiator = reader.int32();
          break;
        case 3:
          message.creator = reader.string();
          break;
        case 4:
          message.contractAddr = reader.string();
          break;
        case 5:
          message.priceDenom = reader.string();
          break;
        case 6:
          message.assetDenom = reader.string();
          break;
        case 7:
          message.positionDirection = reader.int32();
          break;
        case 8:
          message.price = reader.string();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object) {
    var _object$initiator, _object$creator, _object$contractAddr2, _object$priceDenom2, _object$assetDenom2, _object$positionDirec2, _object$price2;
    const message = createBaseCancellation();
    message.id = object.id !== undefined && object.id !== null ? Long.fromValue(object.id) : Long.UZERO;
    message.initiator = (_object$initiator = object.initiator) !== null && _object$initiator !== void 0 ? _object$initiator : 0;
    message.creator = (_object$creator = object.creator) !== null && _object$creator !== void 0 ? _object$creator : undefined;
    message.contractAddr = (_object$contractAddr2 = object.contractAddr) !== null && _object$contractAddr2 !== void 0 ? _object$contractAddr2 : "";
    message.priceDenom = (_object$priceDenom2 = object.priceDenom) !== null && _object$priceDenom2 !== void 0 ? _object$priceDenom2 : "";
    message.assetDenom = (_object$assetDenom2 = object.assetDenom) !== null && _object$assetDenom2 !== void 0 ? _object$assetDenom2 : "";
    message.positionDirection = (_object$positionDirec2 = object.positionDirection) !== null && _object$positionDirec2 !== void 0 ? _object$positionDirec2 : 0;
    message.price = (_object$price2 = object.price) !== null && _object$price2 !== void 0 ? _object$price2 : "";
    return message;
  }
};
function createBaseActiveOrders() {
  return {
    ids: []
  };
}
export const ActiveOrders = {
  encode(message, writer = _m0.Writer.create()) {
    writer.uint32(10).fork();
    for (const v of message.ids) {
      writer.uint64(v);
    }
    writer.ldelim();
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseActiveOrders();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if ((tag & 7) === 2) {
            const end2 = reader.uint32() + reader.pos;
            while (reader.pos < end2) {
              message.ids.push(reader.uint64());
            }
          } else {
            message.ids.push(reader.uint64());
          }
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object) {
    var _object$ids;
    const message = createBaseActiveOrders();
    message.ids = ((_object$ids = object.ids) === null || _object$ids === void 0 ? void 0 : _object$ids.map(e => Long.fromValue(e))) || [];
    return message;
  }
};
