// @sei-js/proto@3.1.0/dist/types/codegen/ibc/applications/transfer/v1/tx.amino.js
import { Long, omitDefault } from "../../../../helpers";
export const AminoConverter = {
    "/ibc.applications.transfer.v1.MsgTransfer": {
        aminoType: "cosmos-sdk/MsgTransfer",
        toAmino: ({ sourcePort, sourceChannel, token, sender, receiver, timeoutHeight, timeoutTimestamp }) => {
            return {
                source_port: sourcePort,
                source_channel: sourceChannel,
                token: {
                    denom: token.denom,
                    amount: Long.fromValue(token.amount).toString()
                },
                sender,
                receiver,
                timeout_height: timeoutHeight ? {
                    revision_height: omitDefault(timeoutHeight.revisionHeight)?.toString(),
                    revision_number: omitDefault(timeoutHeight.revisionNumber)?.toString()
                } : {},
                timeout_timestamp: timeoutTimestamp.toString()
            };
        },
        fromAmino: ({ source_port, source_channel, token, sender, receiver, timeout_height, timeout_timestamp }) => {
            return {
                sourcePort: source_port,
                sourceChannel: source_channel,
                token: {
                    denom: token.denom,
                    amount: token.amount
                },
                sender,
                receiver,
                timeoutHeight: timeout_height ? {
                    revisionHeight: Long.fromString(timeout_height.revision_height || "0", true),
                    revisionNumber: Long.fromString(timeout_height.revision_number || "0", true)
                } : undefined,
                timeoutTimestamp: Long.fromString(timeout_timestamp)
            };
        }
    }
};
