// @injectivelabs/sdk-ts@1.14.16/dist/cjs/core/modules/permissions/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateParams = exports.MsgUpdateNamespaceRoles = exports.MsgUpdateNamespace = exports.MsgRevokeNamespaceRoles = exports.MsgDeleteNamespace = exports.MsgCreateNamespace = exports.MsgClaimVoucher = void 0;
const MsgClaimVoucher_1 = __importDefault(require("./msgs/MsgClaimVoucher"));
exports.MsgClaimVoucher = MsgClaimVoucher_1.default;
const MsgCreateNamespace_1 = __importDefault(require("./msgs/MsgCreateNamespace"));
exports.MsgCreateNamespace = MsgCreateNamespace_1.default;
const MsgDeleteNamespace_1 = __importDefault(require("./msgs/MsgDeleteNamespace"));
exports.MsgDeleteNamespace = MsgDeleteNamespace_1.default;
const MsgRevokeNamespaceRoles_1 = __importDefault(require("./msgs/MsgRevokeNamespaceRoles"));
exports.MsgRevokeNamespaceRoles = MsgRevokeNamespaceRoles_1.default;
const MsgUpdateNamespace_1 = __importDefault(require("./msgs/MsgUpdateNamespace"));
exports.MsgUpdateNamespace = MsgUpdateNamespace_1.default;
const MsgUpdateNamespaceRoles_1 = __importDefault(require("./msgs/MsgUpdateNamespaceRoles"));
exports.MsgUpdateNamespaceRoles = MsgUpdateNamespaceRoles_1.default;
const MsgUpdateParams_1 = __importDefault(require("./msgs/MsgUpdateParams"));
exports.MsgUpdateParams = MsgUpdateParams_1.default;
