// .ignored/@tatumio/tatum/dist/src/util/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Constant = void 0;
const dto_1 = require("../dto");
exports.Constant = {
    TATUM_API_URL: {
        V3: 'https://api.tatum.io/v3/',
        V4: 'https://api.tatum.io/v4/',
    },
    NETWORK: {
        ChainMapInverse: {
            [dto_1.AddressEventNotificationChain.ETH]: dto_1.Chain.Ethereum,
            [dto_1.AddressEventNotificationChain.SOL]: dto_1.Chain.Solana,
            [dto_1.AddressEventNotificationChain.MATIC]: dto_1.Chain.Polygon,
            [dto_1.AddressEventNotificationChain.CELO]: dto_1.Chain.Celo,
            [dto_1.AddressEventNotificationChain.KLAY]: dto_1.Chain.Klaytn,
            [dto_1.AddressEventNotificationChain.BTC]: dto_1.Chain.Bitcoin,
            [dto_1.AddressEventNotificationChain.LTC]: dto_1.Chain.Litecoin,
            [dto_1.AddressEventNotificationChain.BCH]: dto_1.Chain.BitcoinCash,
            [dto_1.AddressEventNotificationChain.DOGE]: dto_1.Chain.Dogecoin,
            [dto_1.AddressEventNotificationChain.TRON]: dto_1.Chain.Tron,
            [dto_1.AddressEventNotificationChain.BSC]: dto_1.Chain.BinanceSmartChain,
            [dto_1.AddressEventNotificationChain.FLR]: dto_1.Chain.Flare,
        },
    },
    DECIMALS: {
        [dto_1.Network.BITCOIN]: 8,
        [dto_1.Network.BITCOIN_TESTNET]: 8,
        [dto_1.Network.MULTIVERSX]: 18,
        [dto_1.Network.MULTIVERSX_TESTNET]: 18,
        [dto_1.Network.NEAR]: 24,
        [dto_1.Network.NEAR_TESTNET]: 24,
        [dto_1.Network.BITCOIN_CASH]: 8,
        [dto_1.Network.ROSTRUM]: 8,
        [dto_1.Network.BITCOIN_CASH_TESTNET]: 8,
        [dto_1.Network.LITECOIN]: 8,
        [dto_1.Network.LITECOIN_TESTNET]: 8,
        [dto_1.Network.DOGECOIN]: 8,
        [dto_1.Network.DOGECOIN_TESTNET]: 8,
        [dto_1.Network.ZCASH]: 8,
        [dto_1.Network.ZCASH_TESTNET]: 8,
        [dto_1.Network.ETHEREUM]: 18,
        [dto_1.Network.ETHEREUM_SEPOLIA]: 18,
        [dto_1.Network.ETHEREUM_HOLESKY]: 18,
        [dto_1.Network.POLYGON]: 18,
        [dto_1.Network.POLYGON_AMOY]: 18,
        [dto_1.Network.CELO]: 18,
        [dto_1.Network.CELO_ALFAJORES]: 18,
        [dto_1.Network.SOLANA]: 9,
        [dto_1.Network.SOLANA_DEVNET]: 9,
        [dto_1.Network.XRP]: 6,
        [dto_1.Network.XRP_TESTNET]: 6,
        [dto_1.Network.KLAYTN]: 18,
        [dto_1.Network.KLAYTN_BAOBAB]: 18,
        [dto_1.Network.TRON]: 6,
        [dto_1.Network.TRON_SHASTA]: 6,
        [dto_1.Network.BINANCE_SMART_CHAIN]: 18,
        [dto_1.Network.BINANCE_SMART_CHAIN_TESTNET]: 18,
        [dto_1.Network.AVALANCHE_C]: 18,
        [dto_1.Network.AVALANCHE_C_TESTNET]: 18,
        [dto_1.Network.AVALANCHE_P]: 18,
        [dto_1.Network.AVALANCHE_P_TESTNET]: 18,
        [dto_1.Network.AVALANCHE_X]: 18,
        [dto_1.Network.AVALANCHE_X_TESTNET]: 18,
        [dto_1.Network.FANTOM]: 18,
        [dto_1.Network.FANTOM_TESTNET]: 18,
        [dto_1.Network.HARMONY_ONE_SHARD_0]: 18,
        [dto_1.Network.HARMONY_ONE_TESTNET_SHARD_0]: 18,
        [dto_1.Network.ALGORAND_ALGOD]: 6,
        [dto_1.Network.ALGORAND_ALGOD_TESTNET]: 6,
        [dto_1.Network.ALGORAND_INDEXER]: 6,
        [dto_1.Network.ALGORAND_INDEXER_TESTNET]: 6,
        [dto_1.Network.ARBITRUM_ONE]: 18,
        [dto_1.Network.ARBITRUM_NOVA_TESTNET]: 18,
        [dto_1.Network.ARBITRUM_NOVA]: 18,
        [dto_1.Network.AURORA]: 18,
        [dto_1.Network.AURORA_TESTNET]: 18,
        [dto_1.Network.CARDANO_ROSETTA]: 6,
        [dto_1.Network.CARDANO_ROSETTA_PREPROD]: 6,
        [dto_1.Network.GNOSIS]: 18,
        [dto_1.Network.GNOSIS_TESTNET]: 18,
        [dto_1.Network.FLOW]: 8,
        [dto_1.Network.FLOW_TESTNET]: 8,
        [dto_1.Network.CRONOS]: 18,
        [dto_1.Network.CRONOS_TESTNET]: 18,
        [dto_1.Network.KUCOIN]: 18,
        [dto_1.Network.KUCOIN_TESTNET]: 18,
        [dto_1.Network.ETHEREUM_CLASSIC]: 18,
        [dto_1.Network.EOS]: 4,
        [dto_1.Network.HORIZEN_EON]: 18,
        [dto_1.Network.HORIZEN_EON_GOBI]: 18,
        [dto_1.Network.EOS_TESTNET]: 4,
        [dto_1.Network.TEZOS]: 6,
        [dto_1.Network.TEZOS_TESTNET]: 6,
        [dto_1.Network.STELLAR]: 6,
        [dto_1.Network.STELLAR_TESTNET]: 6,
        [dto_1.Network.OASIS]: 18,
        [dto_1.Network.OASIS_TESTNET]: 18,
        [dto_1.Network.OPTIMISM]: 18,
        [dto_1.Network.OPTIMISM_TESTNET]: 18,
        [dto_1.Network.PALM]: 18,
        [dto_1.Network.PALM_TESTNET]: 18,
        [dto_1.Network.POLKADOT]: 18,
        [dto_1.Network.POLKADOT_TESTNET]: 18,
        [dto_1.Network.RSK]: 18,
        [dto_1.Network.RSK_TESTNET]: 18,
        [dto_1.Network.VECHAIN]: 18,
        [dto_1.Network.VECHAIN_TESTNET]: 18,
        [dto_1.Network.XINFIN]: 18,
        [dto_1.Network.XINFIN_TESTNET]: 18,
        [dto_1.Network.ZILLIQA]: 12,
        [dto_1.Network.ZILLIQA_TESTNET]: 12,
        [dto_1.Network.FLARE]: 18,
        [dto_1.Network.FLARE_COSTON]: 18,
        [dto_1.Network.FLARE_COSTON_2]: 18,
        [dto_1.Network.FLARE_SONGBIRD]: 18,
        [dto_1.Network.HAQQ]: 18,
        [dto_1.Network.HAQQ_TESTNET]: 18,
        [dto_1.Network.CHILIZ]: 18,
        [dto_1.Network.BNB]: 18,
        [dto_1.Network.BASE]: 18,
        [dto_1.Network.BASE_SEPOLIA]: 18,
        [dto_1.Network.KADENA]: 18,
        [dto_1.Network.KADENA_TESTNET]: 18,
        [dto_1.Network.COSMOS_ROSETTA]: 18,
        [dto_1.Network.IOTA]: 18,
        [dto_1.Network.IOTA_TESTNET]: 18,
        [dto_1.Network.BITCOIN_ELECTRS]: 18,
        [dto_1.Network.BITCOIN_ELECTRS_TESTNET]: 18,
        [dto_1.Network.ROSTRUM_TESTNET]: 18,
        [dto_1.Network.CASPER]: 18,
        [dto_1.Network.TON]: 9,
        [dto_1.Network.TON_TESTNET]: 9,
        [dto_1.Network.ZK_SYNC]: 18,
        [dto_1.Network.ZK_SYNC_TESTNET]: 18,
    },
    CURRENCY_NAMES: {
        [dto_1.Network.BITCOIN]: 'BTC',
        [dto_1.Network.BITCOIN_TESTNET]: 'BTC',
        [dto_1.Network.NEAR]: 'NEAR',
        [dto_1.Network.NEAR_TESTNET]: 'NEAR',
        [dto_1.Network.MULTIVERSX]: 'EGLD',
        [dto_1.Network.MULTIVERSX_TESTNET]: 'EGLD',
        [dto_1.Network.BITCOIN_CASH]: 'BCH',
        [dto_1.Network.BITCOIN_CASH_TESTNET]: 'BCH',
        [dto_1.Network.LITECOIN]: 'LTC',
        [dto_1.Network.LITECOIN_TESTNET]: 'LTC',
        [dto_1.Network.DOGECOIN]: 'DOGE',
        [dto_1.Network.DOGECOIN_TESTNET]: 'DOGE',
        [dto_1.Network.ZCASH]: 'ZEC',
        [dto_1.Network.ZCASH_TESTNET]: 'ZEC',
        [dto_1.Network.ETHEREUM]: 'ETH',
        [dto_1.Network.ETHEREUM_SEPOLIA]: 'ETH',
        [dto_1.Network.ETHEREUM_HOLESKY]: 'ETH',
        [dto_1.Network.POLYGON]: 'MATIC',
        [dto_1.Network.POLYGON_AMOY]: 'MATIC',
        [dto_1.Network.CELO]: 'CELO',
        [dto_1.Network.CELO_ALFAJORES]: 'CELO',
        [dto_1.Network.SOLANA]: 'SOL',
        [dto_1.Network.SOLANA_DEVNET]: 'SOL',
        [dto_1.Network.XRP]: 'XRP',
        [dto_1.Network.XRP_TESTNET]: 'XRP',
        [dto_1.Network.KLAYTN]: 'KLAY',
        [dto_1.Network.KLAYTN_BAOBAB]: 'KLAY',
        [dto_1.Network.TRON]: 'TRX',
        [dto_1.Network.TRON_SHASTA]: 'TRX',
        [dto_1.Network.BINANCE_SMART_CHAIN]: 'BNB',
        [dto_1.Network.BINANCE_SMART_CHAIN_TESTNET]: 'BNB',
        [dto_1.Network.AVALANCHE_C]: 'AVAX',
        [dto_1.Network.AVALANCHE_C_TESTNET]: 'AVAX',
        [dto_1.Network.AVALANCHE_P]: 'AVAX',
        [dto_1.Network.AVALANCHE_P_TESTNET]: 'AVAX',
        [dto_1.Network.AVALANCHE_X]: 'AVAX',
        [dto_1.Network.AVALANCHE_X_TESTNET]: 'AVAX',
        [dto_1.Network.FANTOM]: 'FTM',
        [dto_1.Network.FANTOM_TESTNET]: 'FTM',
        [dto_1.Network.HARMONY_ONE_SHARD_0]: 'ONE',
        [dto_1.Network.HARMONY_ONE_TESTNET_SHARD_0]: 'ONE',
        [dto_1.Network.ALGORAND_ALGOD]: 'ALGO',
        [dto_1.Network.ALGORAND_ALGOD_TESTNET]: 'ALGO',
        [dto_1.Network.ALGORAND_INDEXER]: 'ALGO',
        [dto_1.Network.ALGORAND_INDEXER_TESTNET]: 'ALGO',
        [dto_1.Network.ARBITRUM_ONE]: 'ARB',
        [dto_1.Network.ARBITRUM_NOVA_TESTNET]: 'ARB',
        [dto_1.Network.ARBITRUM_NOVA]: 'ARB',
        [dto_1.Network.AURORA]: 'AURA',
        [dto_1.Network.AURORA_TESTNET]: 'AURA',
        [dto_1.Network.CARDANO_ROSETTA]: 'ADA',
        [dto_1.Network.CARDANO_ROSETTA_PREPROD]: 'ADA',
        [dto_1.Network.GNOSIS]: 'GNO',
        [dto_1.Network.GNOSIS_TESTNET]: 'GNO',
        [dto_1.Network.FLOW]: 'FLOW',
        [dto_1.Network.FLARE]: 'FLARE',
        [dto_1.Network.FLARE_COSTON]: 'FLARE',
        [dto_1.Network.FLARE_COSTON_2]: 'FLARE',
        [dto_1.Network.FLARE_SONGBIRD]: 'FLARE',
        [dto_1.Network.FLOW_TESTNET]: 'FLOW',
        [dto_1.Network.CRONOS]: 'CRO',
        [dto_1.Network.CRONOS_TESTNET]: 'CRO',
        [dto_1.Network.KUCOIN]: 'KCS',
        [dto_1.Network.KUCOIN_TESTNET]: 'KCS',
        [dto_1.Network.ETHEREUM_CLASSIC]: 'ETC',
        [dto_1.Network.EOS]: 'EOS',
        [dto_1.Network.HORIZEN_EON]: 'EON',
        [dto_1.Network.HORIZEN_EON_GOBI]: 'EON',
        [dto_1.Network.EOS_TESTNET]: 'EOS',
        [dto_1.Network.TEZOS]: 'XTZ',
        [dto_1.Network.TEZOS_TESTNET]: 'XTZ',
        [dto_1.Network.STELLAR]: 'XLM',
        [dto_1.Network.STELLAR_TESTNET]: 'XLM',
        [dto_1.Network.OASIS]: 'ROSE',
        [dto_1.Network.OASIS_TESTNET]: 'ROSE',
        [dto_1.Network.OPTIMISM]: 'OPT',
        [dto_1.Network.OPTIMISM_TESTNET]: 'OPT',
        [dto_1.Network.PALM]: 'PALM',
        [dto_1.Network.PALM_TESTNET]: 'PALM',
        [dto_1.Network.POLKADOT]: 'DOT',
        [dto_1.Network.POLKADOT_TESTNET]: 'DOT',
        [dto_1.Network.RSK]: 'RBTC',
        [dto_1.Network.RSK_TESTNET]: 'RBTC',
        [dto_1.Network.VECHAIN]: 'VET',
        [dto_1.Network.VECHAIN_TESTNET]: 'VET',
        [dto_1.Network.XINFIN]: 'XDC',
        [dto_1.Network.XINFIN_TESTNET]: 'XDC',
        [dto_1.Network.ZILLIQA]: 'ZIL',
        [dto_1.Network.ZILLIQA_TESTNET]: 'ZIL',
        [dto_1.Network.HAQQ]: 'HAQQ',
        [dto_1.Network.HAQQ_TESTNET]: 'HAQQ',
        [dto_1.Network.CHILIZ]: 'CHILIZ',
        [dto_1.Network.BNB]: 'BNB',
        [dto_1.Network.BASE]: 'BASE',
        [dto_1.Network.BASE_SEPOLIA]: 'BASE',
        [dto_1.Network.KADENA]: 'KADENA',
        [dto_1.Network.KADENA_TESTNET]: 'KADENA',
        [dto_1.Network.ROSTRUM]: 'BCH',
        [dto_1.Network.COSMOS_ROSETTA]: 'ATOM',
        [dto_1.Network.IOTA]: 'IOTA',
        [dto_1.Network.IOTA_TESTNET]: 'IOTA',
        [dto_1.Network.BITCOIN_ELECTRS]: 'BTC',
        [dto_1.Network.BITCOIN_ELECTRS_TESTNET]: 'BTC',
        [dto_1.Network.ROSTRUM_TESTNET]: 'BCH',
        [dto_1.Network.CASPER]: 'CASPER',
        [dto_1.Network.TON]: 'TON',
        [dto_1.Network.TON_TESTNET]: 'TON',
        [dto_1.Network.ZK_SYNC]: 'ZKS',
        [dto_1.Network.ZK_SYNC_TESTNET]: 'ZKS',
    },
    RPC: {
        MAINNETS: [
            dto_1.Network.ALGORAND_ALGOD,
            dto_1.Network.ARBITRUM_NOVA,
            dto_1.Network.ARBITRUM_ONE,
            dto_1.Network.AURORA,
            dto_1.Network.AVALANCHE_C,
            dto_1.Network.AVALANCHE_P,
            dto_1.Network.AVALANCHE_X,
            dto_1.Network.BINANCE_SMART_CHAIN,
            dto_1.Network.BITCOIN,
            dto_1.Network.BITCOIN_CASH,
            dto_1.Network.CARDANO_ROSETTA,
            dto_1.Network.CELO,
            dto_1.Network.CRONOS,
            dto_1.Network.DOGECOIN,
            dto_1.Network.EOS,
            dto_1.Network.ETHEREUM,
            dto_1.Network.ETHEREUM_CLASSIC,
            dto_1.Network.FANTOM,
            dto_1.Network.FLOW,
            dto_1.Network.GNOSIS,
            dto_1.Network.HARMONY_ONE_SHARD_0,
            dto_1.Network.KLAYTN,
            dto_1.Network.KUCOIN,
            dto_1.Network.LITECOIN,
            dto_1.Network.MULTIVERSX,
            dto_1.Network.NEAR,
            dto_1.Network.OASIS,
            dto_1.Network.OPTIMISM,
            dto_1.Network.PALM,
            dto_1.Network.POLYGON,
            dto_1.Network.POLKADOT,
            dto_1.Network.RSK,
            dto_1.Network.SOLANA,
            dto_1.Network.STELLAR,
            dto_1.Network.TEZOS,
            dto_1.Network.TRON,
            dto_1.Network.VECHAIN,
            dto_1.Network.XINFIN,
            dto_1.Network.XRP,
            dto_1.Network.ZCASH,
            dto_1.Network.ZILLIQA,
            dto_1.Network.CHILIZ,
        ],
        TESTNETS: [
            dto_1.Network.ALGORAND_ALGOD_TESTNET,
            dto_1.Network.ARBITRUM_NOVA_TESTNET,
            dto_1.Network.AURORA_TESTNET,
            dto_1.Network.AVALANCHE_C_TESTNET,
            dto_1.Network.AVALANCHE_P_TESTNET,
            dto_1.Network.AVALANCHE_X_TESTNET,
            dto_1.Network.BINANCE_SMART_CHAIN_TESTNET,
            dto_1.Network.BITCOIN_TESTNET,
            dto_1.Network.BITCOIN_CASH_TESTNET,
            dto_1.Network.CARDANO_ROSETTA_PREPROD,
            dto_1.Network.CELO_ALFAJORES,
            dto_1.Network.CRONOS_TESTNET,
            dto_1.Network.DOGECOIN_TESTNET,
            dto_1.Network.ETHEREUM_SEPOLIA,
            dto_1.Network.ETHEREUM_HOLESKY,
            dto_1.Network.EOS_TESTNET,
            dto_1.Network.FANTOM_TESTNET,
            dto_1.Network.FLOW_TESTNET,
            dto_1.Network.GNOSIS_TESTNET,
            dto_1.Network.HARMONY_ONE_TESTNET_SHARD_0,
            dto_1.Network.KLAYTN_BAOBAB,
            dto_1.Network.KUCOIN_TESTNET,
            dto_1.Network.LITECOIN_TESTNET,
            dto_1.Network.MULTIVERSX_TESTNET,
            dto_1.Network.NEAR_TESTNET,
            dto_1.Network.OASIS_TESTNET,
            dto_1.Network.OPTIMISM_TESTNET,
            dto_1.Network.PALM_TESTNET,
            dto_1.Network.POLYGON_AMOY,
            dto_1.Network.POLKADOT_TESTNET,
            dto_1.Network.RSK_TESTNET,
            dto_1.Network.SOLANA_DEVNET,
            dto_1.Network.TEZOS_TESTNET,
            dto_1.Network.TRON_SHASTA,
            dto_1.Network.VECHAIN_TESTNET,
            dto_1.Network.XINFIN_TESTNET,
            dto_1.Network.XRP_TESTNET,
            dto_1.Network.ZCASH_TESTNET,
            dto_1.Network.ZILLIQA_TESTNET,
            dto_1.Network.HORIZEN_EON_GOBI,
        ],
        METHOD_PREFIX: 'eth_',
    },
    OPEN_RPC: {
        LB_INTERVAL: 60000,
        ALLOWED_BLOCKS_BEHIND: 10,
    },
    TRON_SHASTA_BASE_URL: {
        BASE: 'https://api.shasta.trongrid.io',
        RPC: 'https://api.shasta.trongrid.io/jsonrpc',
    },
    EOS_PREFIX: '/v1/chain/',
    BEACON_PREFIX: '/eth/v1/beacon',
    BEACON_BASE_PREFIX: '/eth/v1',
};
