// @sei-js/proto@3.1.0/dist/types/codegen/seiprotocol/seichain/tokenfactory/tx.amino.js
import { Long } from "../../../helpers";
export const AminoConverter = {
    "/seiprotocol.seichain.tokenfactory.MsgCreateDenom": {
        aminoType: "/seiprotocol.seichain.tokenfactory.MsgCreateDenom",
        toAmino: ({ sender, subdenom }) => {
            return {
                sender,
                subdenom
            };
        },
        fromAmino: ({ sender, subdenom }) => {
            return {
                sender,
                subdenom
            };
        }
    },
    "/seiprotocol.seichain.tokenfactory.MsgMint": {
        aminoType: "/seiprotocol.seichain.tokenfactory.MsgMint",
        toAmino: ({ sender, amount }) => {
            return {
                sender,
                amount: {
                    denom: amount.denom,
                    amount: Long.fromValue(amount.amount).toString()
                }
            };
        },
        fromAmino: ({ sender, amount }) => {
            return {
                sender,
                amount: {
                    denom: amount.denom,
                    amount: amount.amount
                }
            };
        }
    },
    "/seiprotocol.seichain.tokenfactory.MsgBurn": {
        aminoType: "/seiprotocol.seichain.tokenfactory.MsgBurn",
        toAmino: ({ sender, amount }) => {
            return {
                sender,
                amount: {
                    denom: amount.denom,
                    amount: Long.fromValue(amount.amount).toString()
                }
            };
        },
        fromAmino: ({ sender, amount }) => {
            return {
                sender,
                amount: {
                    denom: amount.denom,
                    amount: amount.amount
                }
            };
        }
    },
    "/seiprotocol.seichain.tokenfactory.MsgChangeAdmin": {
        aminoType: "/seiprotocol.seichain.tokenfactory.MsgChangeAdmin",
        toAmino: ({ sender, denom, newAdmin }) => {
            return {
                sender,
                denom,
                new_admin: newAdmin
            };
        },
        fromAmino: ({ sender, denom, new_admin }) => {
            return {
                sender,
                denom,
                newAdmin: new_admin
            };
        }
    }
};
